File: Common/JwtModules/GoJwt/JwtReader.go
package jwtreader

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"regexp"
	"strings"
	"time"
)

var (
	ErrInvalidTokenFormat = errors.New("invalid JWT format")
	ErrInvalidAlgorithm   = errors.New("unsupported algorithm; expected HS256")
	ErrInvalidSignature   = errors.New("invalid JWT signature")
	ErrTokenExpired       = errors.New("token expired")
	ErrTokenNotYetValid   = errors.New("token not valid yet")
	ErrMissingClaim       = errors.New("missing required claim")
	ErrInvalidClaimType   = errors.New("invalid claim type")
	ErrInvalidRole        = errors.New("invalid role")
	ErrInvalidUsername    = errors.New("invalid username")
)

var usernameRe = regexp.MustCompile(`^[a-zA-Z0-9_]{3,30}$`)

type Claims struct {
	UserID   int64  `json:"user_id"`
	Username string `json:"username"`
	Email    string `json:"email"`
	Role     string `json:"role"`

	Sub int64  `json:"sub"`
	Iat int64  `json:"iat"`
	Exp int64  `json:"exp"`
	Nbf int64  `json:"nbf"`
	Jti string `json:"jti,omitempty"`
}

type Header struct {
	Alg string `json:"alg"`
	Typ string `json:"typ"`
}

type Token struct {
	Header  Header
	Claims  Claims
	Raw     string
	Subject int64
}

func ExtractBearerToken(authHeader string) (string, error) {
	if !strings.HasPrefix(authHeader, "Bearer ") {
		return "", errors.New("authorization header must start with Bearer ")
	}
	token := strings.TrimSpace(strings.TrimPrefix(authHeader, "Bearer "))
	if token == "" {
		return "", errors.New("empty bearer token")
	}
	return token, nil
}

func ReadFromAuthorizationHeader(authHeader string) (*Token, error) {
	token, err := ExtractBearerToken(authHeader)
	if err != nil {
		return nil, err
	}
	return ParseAndValidate(token, os.Getenv("JWT_SECRET"), time.Now().Unix())
}

func ParseAndValidate(jwt string, secret string, nowUnix int64) (*Token, error) {
	if secret == "" {
		return nil, errors.New("JWT_SECRET is empty")
	}

	parts := strings.Split(jwt, ".")
	if len(parts) != 3 {
		return nil, ErrInvalidTokenFormat
	}

	headerJSON, err := base64.RawURLEncoding.DecodeString(parts[0])
	if err != nil {
		return nil, fmt.Errorf("decode header: %w", err)
	}

	payloadJSON, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return nil, fmt.Errorf("decode payload: %w", err)
	}

	var header Header
	if err := json.Unmarshal(headerJSON, &header); err != nil {
		return nil, fmt.Errorf("unmarshal header: %w", err)
	}
	if header.Alg != "HS256" {
		return nil, ErrInvalidAlgorithm
	}

	signingInput := parts[0] + "." + parts[1]
	if !verifyHS256(signingInput, parts[2], secret) {
		return nil, ErrInvalidSignature
	}

	var claims Claims
	if err := json.Unmarshal(payloadJSON, &claims); err != nil {
		return nil, fmt.Errorf("unmarshal claims: %w", err)
	}

	if err := validateClaims(claims, nowUnix); err != nil {
		return nil, err
	}

	return &Token{
		Header:  header,
		Claims:  claims,
		Raw:     jwt,
		Subject: claims.Sub,
	}, nil
}

func verifyHS256(signingInput, signatureB64 string, secret string) bool {
	mac := hmac.New(sha256.New, []byte(secret))
	_, _ = mac.Write([]byte(signingInput))
	expected := mac.Sum(nil)

	got, err := base64.RawURLEncoding.DecodeString(signatureB64)
	if err != nil {
		return false
	}
	return hmac.Equal(expected, got)
}

func validateClaims(c Claims, nowUnix int64) error {
	if c.UserID == 0 {
		return ErrMissingClaim
	}
	if c.Username == "" {

		return ErrMissingClaim
	}
	if c.Email == "" {
		return ErrMissingClaim
	}
	if c.Role == "" {
		return ErrMissingClaim
	}
	if c.Sub == 0 {
		return ErrMissingClaim
	}
	if c.Iat == 0 || c.Exp == 0 || c.Nbf == 0 {
		return ErrMissingClaim
	}

	if c.Role != "user" && c.Role != "admin" {
		return ErrInvalidRole
	}
	if !usernameRe.MatchString(c.Username) {
		return ErrInvalidUsername
	}
	if c.Exp < nowUnix {
		return ErrTokenExpired
	}
	if c.Nbf > nowUnix {
		return ErrTokenNotYetValid
	}
	if c.UserID != c.Sub {
		return errors.New("user_id must match sub")
	}

	return nil
}

File: Common/JwtModules/PythonJwt/JwtReader.py
import base64
import hashlib
import hmac
import json
import os
import re
import time
from dataclasses import dataclass
from typing import Any, Dict, Optional


USERNAME_RE = re.compile(r"^[a-zA-Z0-9_]{3,30}$")


class JwtError(Exception):
    pass


class InvalidTokenFormat(JwtError):
    pass


class InvalidAlgorithm(JwtError):
    pass


class InvalidSignature(JwtError):
    pass


class TokenExpired(JwtError):
    pass


class TokenNotYetValid(JwtError):
    pass


class MissingClaim(JwtError):
    pass


class InvalidRole(JwtError):
    pass


class InvalidUsername(JwtError):
    pass


def extract_bearer_token(authorization_header: str) -> str:
    if not authorization_header or not authorization_header.startswith("Bearer "):
        raise ValueError("Authorization header must start with Bearer ")
    token = authorization_header[len("Bearer "):].strip()
    if not token:
        raise ValueError("Empty bearer token")
    return token


def read_from_authorization_header(authorization_header: str, secret: Optional[str] = None) -> Dict[str, Any]:
    token = extract_bearer_token(authorization_header)
    return parse_and_validate(token, secret or os.getenv("JWT_SECRET", ""), int(time.time()))


def parse_and_validate(jwt: str, secret: str, now_unix: int) -> Dict[str, Any]:
    if not secret:
        raise ValueError("JWT secret is empty")

    parts = jwt.split(".")
    if len(parts) != 3:
        raise InvalidTokenFormat("Invalid JWT format")

    header = json.loads(_b64url_decode(parts[0]))
    payload = json.loads(_b64url_decode(parts[1]))

    if header.get("alg") != "HS256":
        raise InvalidAlgorithm("Unsupported algorithm; expected HS256")

    signing_input = f"{parts[0]}.{parts[1]}".encode("utf-8")
    if not _verify_hs256(signing_input, parts[2], secret):
        raise InvalidSignature("Invalid JWT signature")

    _validate_claims(payload, now_unix)
    return payload


def _verify_hs256(signing_input: bytes, signature_b64url: str, secret: str) -> bool:
    expected = hmac.new(secret.encode("utf-8"), signing_input, hashlib.sha256).digest()
    try:
        got = _b64url_decode_bytes(signature_b64url)
    except Exception:
        return False
    return hmac.compare_digest(expected, got)


def _validate_claims(c: Dict[str, Any], now_unix: int) -> None:
    required = ["user_id", "username", "email", "role", "sub", "iat", "exp", "nbf"]
    for key in required:
        if key not in c or c[key] in (None, ""):
            raise MissingClaim(f"Missing required claim: {key}")

    if c["role"] not in ("user", "admin"):
        raise InvalidRole("Invalid role")

    if not isinstance(c["username"], str) or not USERNAME_RE.fullmatch(c["username"]):
        raise InvalidUsername("Invalid username")

    if not isinstance(c["user_id"], int) or not isinstance(c["sub"], str):
        raise JwtError("user_id must be integer, sub must be string")

    try:
        if c["user_id"] != int(c["sub"]):
            raise JwtError("user_id must match sub")
    except ValueError:
        raise JwtError("sub must be a numeric string")

    if not isinstance(c["iat"], int) or not isinstance(c["exp"], int) or not isinstance(c["nbf"], int):
        raise JwtError("Timestamp claims must be integers")

    if c["exp"] < now_unix:
        raise TokenExpired("Token expired")
    if c["nbf"] > now_unix:
        raise TokenNotYetValid("Token not yet valid")
    if c["user_id"] != c["sub"]:
        raise JwtError("user_id must match sub")


def _b64url_decode(text: str) -> str:
    return _b64url_decode_bytes(text).decode("utf-8")


def _b64url_decode_bytes(text: str) -> bytes:
    padding = "=" * (-len(text) % 4)
    return base64.urlsafe_b64decode(text + padding)
